File: python_pure/src/volarb_py/svi.py
from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Final, Literal

SviStatus = Literal[
    "arbitrage_free_on_grid",
    "butterfly_arbitrage_found",
    "invalid_parameters",
]

MINIMUM_TOTAL_VARIANCE: Final[float] = 1e-12
DEFAULT_SCAN_STEPS: Final[int] = 512
MINIMUM_SCAN_STEPS: Final[int] = 2
REFINEMENT_PASSES: Final[int] = 60
GOLDEN_SECTION_RATIO: Final[float] = 0.6180339887498949
BUTTERFLY_TOLERANCE: Final[float] = -1e-12


class InvalidSviParametersError(ValueError):
    pass


@dataclass(frozen=True)
class SviParameters:
    a: float
    b: float
    rho: float
    m: float
    sigma: float


@dataclass(frozen=True)
class SviSliceScan:
    minimum_durrleman_value: float
    log_moneyness_at_minimum: float
    minimum_total_variance: float
    minimum_risk_neutral_density: float
    scan_steps: int
    status: SviStatus


def validate_svi_parameters(parameters: SviParameters) -> None:
    if parameters.b < 0.0:
        raise InvalidSviParametersError(f"b must not be negative, got {parameters.b}")
    if not -1.0 < parameters.rho < 1.0:
        raise InvalidSviParametersError(f"rho must lie strictly inside (-1, 1), got {parameters.rho}")
    if parameters.sigma <= 0.0:
        raise InvalidSviParametersError(f"sigma must be positive, got {parameters.sigma}")
    floor = parameters.a + parameters.b * parameters.sigma * math.sqrt(1.0 - parameters.rho * parameters.rho)
    if floor < 0.0:
        raise InvalidSviParametersError(f"minimum total variance must not be negative, got {floor}")


def total_variance(parameters: SviParameters, log_moneyness: float) -> float:
    centred = log_moneyness - parameters.m
    root = math.sqrt(centred * centred + parameters.sigma * parameters.sigma)
    return parameters.a + parameters.b * (parameters.rho * centred + root)


def total_variance_first_derivative(parameters: SviParameters, log_moneyness: float) -> float:
    centred = log_moneyness - parameters.m
    root = math.sqrt(centred * centred + parameters.sigma * parameters.sigma)
    return parameters.b * (parameters.rho + centred / root)


def total_variance_second_derivative(parameters: SviParameters, log_moneyness: float) -> float:
    centred = log_moneyness - parameters.m
    root = math.sqrt(centred * centred + parameters.sigma * parameters.sigma)
    return parameters.b * parameters.sigma * parameters.sigma / (root * root * root)


def implied_volatility(parameters: SviParameters, log_moneyness: float, years_to_expiry: float) -> float:
    return math.sqrt(max(total_variance(parameters, log_moneyness), 0.0) / years_to_expiry)


def durrleman_function(parameters: SviParameters, log_moneyness: float) -> float:
    variance = max(total_variance(parameters, log_moneyness), MINIMUM_TOTAL_VARIANCE)
    slope = total_variance_first_derivative(parameters, log_moneyness)
    curvature = total_variance_second_derivative(parameters, log_moneyness)
    balance = 1.0 - log_moneyness * slope / (2.0 * variance)
    return balance * balance - 0.25 * slope * slope * (1.0 / variance + 0.25) + 0.5 * curvature


def risk_neutral_density(parameters: SviParameters, log_moneyness: float) -> float:
    variance = max(total_variance(parameters, log_moneyness), MINIMUM_TOTAL_VARIANCE)
    root = math.sqrt(variance)
    standardized = -log_moneyness / root - 0.5 * root
    weight = math.exp(-0.5 * standardized * standardized) / math.sqrt(2.0 * math.pi * variance)
    return durrleman_function(parameters, log_moneyness) * weight


def refined_minimum(parameters: SviParameters, lower: float, upper: float) -> tuple[float, float]:
    left = upper - GOLDEN_SECTION_RATIO * (upper - lower)
    right = lower + GOLDEN_SECTION_RATIO * (upper - lower)
    left_value = durrleman_function(parameters, left)
    right_value = durrleman_function(parameters, right)

    for _ in range(REFINEMENT_PASSES):
        if left_value < right_value:
            upper, right, right_value = right, left, left_value
            left = upper - GOLDEN_SECTION_RATIO * (upper - lower)
            left_value = durrleman_function(parameters, left)
        else:
            lower, left, left_value = left, right, right_value
            right = lower + GOLDEN_SECTION_RATIO * (upper - lower)
            right_value = durrleman_function(parameters, right)

    if left_value <= right_value:
        return left, left_value
    return right, right_value


def scan_svi_slice(
    parameters: SviParameters,
    lowest_log_moneyness: float,
    highest_log_moneyness: float,
    scan_steps: int = DEFAULT_SCAN_STEPS,
) -> SviSliceScan:
    validate_svi_parameters(parameters)
    if not highest_log_moneyness > lowest_log_moneyness:
        raise InvalidSviParametersError("the scan range must be non-empty and increasing")
    if scan_steps < MINIMUM_SCAN_STEPS:
        raise InvalidSviParametersError(f"scan_steps must be at least {MINIMUM_SCAN_STEPS}, got {scan_steps}")

    span = highest_log_moneyness - lowest_log_moneyness
    grid = [lowest_log_moneyness + span * index / scan_steps for index in range(scan_steps + 1)]
    values = [durrleman_function(parameters, point) for point in grid]

    best_index = min(range(len(values)), key=lambda index: (values[index], index))
    neighbourhood_lower = grid[max(best_index - 1, 0)]
    neighbourhood_upper = grid[min(best_index + 1, scan_steps)]
    refined_point, refined_value = refined_minimum(parameters, neighbourhood_lower, neighbourhood_upper)

    if refined_value <= values[best_index]:
        worst_point, worst_value = refined_point, refined_value
    else:
        worst_point, worst_value = grid[best_index], values[best_index]

    return SviSliceScan(
        minimum_durrleman_value=worst_value,
        log_moneyness_at_minimum=worst_point,
        minimum_total_variance=min(total_variance(parameters, point) for point in grid),
        minimum_risk_neutral_density=risk_neutral_density(parameters, worst_point),
        scan_steps=scan_steps,
        status="arbitrage_free_on_grid"
        if worst_value >= BUTTERFLY_TOLERANCE
        else "butterfly_arbitrage_found",
    )

File: cpp_core/src/svi.cpp
#include "volarb/svi.hpp"

#include <algorithm>
#include <cmath>
#include <numbers>
#include <utility>
#include <vector>

namespace volarb {

namespace {

struct MinimumLocation {
    double log_moneyness;
    double value;
};

MinimumLocation refined_minimum(const SviParameters& parameters, double lower, double upper) {
    double left = upper - golden_section_ratio * (upper - lower);
    double right = lower + golden_section_ratio * (upper - lower);
    double left_value = durrleman_function(parameters, left);
    double right_value = durrleman_function(parameters, right);

    for (int pass = 0; pass < refinement_passes; ++pass) {
        if (left_value < right_value) {
            upper = right;
            right = left;
            right_value = left_value;
            left = upper - golden_section_ratio * (upper - lower);
            left_value = durrleman_function(parameters, left);
        } else {
            lower = left;
            left = right;
            left_value = right_value;
            right = lower + golden_section_ratio * (upper - lower);
            right_value = durrleman_function(parameters, right);
        }
    }

    if (left_value <= right_value) {
        return MinimumLocation{left, left_value};
    }
    return MinimumLocation{right, right_value};
}

}

std::string name_of_svi_status(SviStatus status) {
    switch (status) {
    case SviStatus::ArbitrageFreeOnGrid:
        return "arbitrage_free_on_grid";
    case SviStatus::ButterflyArbitrageFound:
        return "butterfly_arbitrage_found";
    case SviStatus::InvalidParameters:
        return "invalid_parameters";
    }
    return "invalid_parameters";
}

void validate_svi_parameters(const SviParameters& parameters) {
    if (parameters.b < 0.0) {
        throw InvalidSviParametersError("b must not be negative, got " + std::to_string(parameters.b));
    }
    if (!(parameters.rho > -1.0 && parameters.rho < 1.0)) {
        throw InvalidSviParametersError("rho must lie strictly inside (-1, 1), got " +
                                        std::to_string(parameters.rho));
    }
    if (parameters.sigma <= 0.0) {
        throw InvalidSviParametersError("sigma must be positive, got " + std::to_string(parameters.sigma));
    }
    const double floor =
        parameters.a + parameters.b * parameters.sigma * std::sqrt(1.0 - parameters.rho * parameters.rho);
    if (floor < 0.0) {
        throw InvalidSviParametersError("minimum total variance must not be negative, got " +
                                        std::to_string(floor));
    }
}

double total_variance(const SviParameters& parameters, double log_moneyness) {
    const double centred = log_moneyness - parameters.m;
    const double root = std::sqrt(centred * centred + parameters.sigma * parameters.sigma);
    return parameters.a + parameters.b * (parameters.rho * centred + root);
}

double total_variance_first_derivative(const SviParameters& parameters, double log_moneyness) {
    const double centred = log_moneyness - parameters.m;
    const double root = std::sqrt(centred * centred + parameters.sigma * parameters.sigma);
    return parameters.b * (parameters.rho + centred / root);
}

double total_variance_second_derivative(const SviParameters& parameters, double log_moneyness) {
    const double centred = log_moneyness - parameters.m;
    const double root = std::sqrt(centred * centred + parameters.sigma * parameters.sigma);
    return parameters.b * parameters.sigma * parameters.sigma / (root * root * root);
}

double implied_volatility(const SviParameters& parameters, double log_moneyness, double years_to_expiry) {
    return std::sqrt(std::max(total_variance(parameters, log_moneyness), 0.0) / years_to_expiry);
}

double durrleman_value(double log_moneyness, double variance, double slope, double curvature) {
    const double balance = 1.0 - log_moneyness * slope / (2.0 * variance);
    return balance * balance - 0.25 * slope * slope * (1.0 / variance + 0.25) + 0.5 * curvature;
}

double density_weight(double log_moneyness, double variance) {
    const double root = std::sqrt(variance);
    const double standardized = -log_moneyness / root - 0.5 * root;
    return std::exp(-0.5 * standardized * standardized) / std::sqrt(2.0 * std::numbers::pi * variance);
}

double durrleman_function(const SviParameters& parameters, double log_moneyness) {
    const double variance = std::max(total_variance(parameters, log_moneyness), minimum_total_variance);
    const double slope = total_variance_first_derivative(parameters, log_moneyness);
    const double curvature = total_variance_second_derivative(parameters, log_moneyness);
    return durrleman_value(log_moneyness, variance, slope, curvature);
}

double risk_neutral_density(const SviParameters& parameters, double log_moneyness) {
    const double variance = std::max(total_variance(parameters, log_moneyness), minimum_total_variance);
    return durrleman_function(parameters, log_moneyness) * density_weight(log_moneyness, variance);
}

SviSliceScan scan_svi_slice(const SviParameters& parameters, double lowest_log_moneyness,
                            double highest_log_moneyness, int scan_steps) {
    validate_svi_parameters(parameters);
    if (!(highest_log_moneyness > lowest_log_moneyness)) {
        throw InvalidSviParametersError("the scan range must be non-empty and increasing");
    }
    if (scan_steps < minimum_scan_steps) {
        throw InvalidSviParametersError("scan_steps must be at least " +
                                        std::to_string(minimum_scan_steps) + ", got " +
                                        std::to_string(scan_steps));
    }

    const double span = highest_log_moneyness - lowest_log_moneyness;
    std::vector<double> grid;
    std::vector<double> values;
    grid.reserve(static_cast<std::size_t>(scan_steps + 1));
    values.reserve(static_cast<std::size_t>(scan_steps + 1));
    for (int index = 0; index <= scan_steps; ++index) {
        const double point =
            lowest_log_moneyness + span * static_cast<double>(index) / static_cast<double>(scan_steps);
        grid.push_back(point);
        values.push_back(durrleman_function(parameters, point));
    }

    std::size_t best_index = 0;
    for (std::size_t index = 1; index < values.size(); ++index) {
        if (values[index] < values[best_index]) {
            best_index = index;
        }
    }

    const double neighbourhood_lower = grid[best_index == 0 ? 0 : best_index - 1];
    const double neighbourhood_upper =
        grid[std::min(best_index + 1, static_cast<std::size_t>(scan_steps))];
    const MinimumLocation refined = refined_minimum(parameters, neighbourhood_lower, neighbourhood_upper);

    double worst_point = grid[best_index];
    double worst_value = values[best_index];
    if (refined.value <= worst_value) {
        worst_point = refined.log_moneyness;
        worst_value = refined.value;
    }

    double smallest_total_variance = total_variance(parameters, grid[0]);
    for (const double point : grid) {
        smallest_total_variance = std::min(smallest_total_variance, total_variance(parameters, point));
    }

    return SviSliceScan{
        worst_value,
        worst_point,
        smallest_total_variance,
        risk_neutral_density(parameters, worst_point),
        scan_steps,
        worst_value >= butterfly_tolerance ? SviStatus::ArbitrageFreeOnGrid
                                           : SviStatus::ButterflyArbitrageFound,
    };
}

}
